src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,224 @@
+                ->update($this->configuration['table_name'])
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id');
+            $this->executeQuery($sql, [
+                ':id' => $doctrineEnvelope['id'],
+                ':delivered_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+            ]);
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Throwable $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+    }
+
+    public function reject($id): bool
+    {
+        return $this->driverConnection->update($this->configuration['table_name'], ['delivered_at' => null], ['id' => $id]) > 0;
+    }
+
+    public function requeueMessages()
+    {
+        $redeliverLimit = new \DateTime('now');
+        $redeliverLimit->modify(sprintf('-%d second', $this->configuration['redeliver_timeout']));
+        $sql = $this->driverConnection->createQueryBuilder()
+            ->update($this->configuration['table_name'])
+            ->set('delivered_at', ':delivered_at')
+            ->andWhere('delivered_at < :redeliver_limit')
+            ->andWhere('queue = :queue')
+            ->getSQL();
+
+        $this->executeQuery($sql, [
+            ':delivered_at' => null,
+            ':queue' => $this->configuration['queue'],
+            ':redeliver_limit' => $redeliverLimit->format('Y-m-d H:i:s'),
+        ]);
+    }
+
+    private function executeQuery(string $sql, array $parameters = [])
+    {
+        $stmt = null;
+        try {
+            $stmt = $this->driverConnection->prepare($sql);
+            $stmt->execute($parameters);
+        } catch (TableNotFoundException $e) {
+            // create table
+            if (!$this->driverConnection->isTransactionActive() && $this->configuration['auto_setup']) {
+                $this->setup();
+            }
+            // statement not prepared ? SQLite throw on exception on prepare if the table does not exist
+            if (null === $stmt) {
+                $stmt = $this->driverConnection->prepare($sql);
+            }
+            $stmt->execute($parameters);
+        }
+
+        return $stmt;
+    }
+
+    private function setup(): void
+    {
+        $synchronizer = new SingleDatabaseSynchronizer($this->driverConnection);
+        $synchronizer->updateSchema($this->getSchema(), true);
+    }
+
+    private function getSchema(): Schema
+    {
+        $schema = new Schema();
+        $table = $schema->createTable($this->configuration['table_name']);
+        $table->addColumn('id', Type::BIGINT)
+            ->setAutoincrement(true)
+            ->setNotnull(true);
+        $table->addColumn('body', Type::TEXT)
+            ->setNotnull(true);
+        $table->addColumn('headers', Type::TEXT)
```

Comment:
I think I disagree - it feels very weird to me to be manually handling a JSON field, when that exists inside of Doctrine already and has built-in exceptions if there is a decoding problem.

Reply:
Just pushed the modification to use a `json` column. I had to call the `Doctrine\DBAL\Types\JsonType::convertToDatabaseValue` and `Doctrine\DBAL\Types\JsonType::convertToPHPValue`.